### PR TITLE
Feature/extend policy

### PR DESCRIPTION
**Pull request by yusank**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -102,6 +102,14 @@ fmt-check: ## Check project files format info
 vet: ## Vet project files
 	@$(GO) vet $(VETPACKAGES)
 
+.PHONY: docs
+docs-generate: ## generate crd api docs
```
Comment (likakuli) [inline Makefile]: add dependency just as controller-gen

File: apis/policy/v1alpha1/clustervalidatepolicy_types.go [hunk 2]
```diff
@@ -42,8 +42,67 @@ type ValidateRuleWithOperation struct {
 	TargetOperations []admissionv1.Operation `json:"targetOperations,omitempty"`
 
 	// Cue represents validate rules defined with cue code.
-	// +required
+	// +optional
 	Cue string `json:"cue"`
+
+	// Template of condition which defines validate cond, and
+	// it will be rendered to CUE and store in RenderedCue field, so
+	//if there are any data added manually will be erased.
```
Comment (likakuli) [inline apis/policy/v1alpha1/clustervalidatepolicy_types.go]: set first blank space

**Comment by yusank** [comment]
doc url: [https://k-cloud-labs.github.io/kinitiras-doc/](https://k-cloud-labs.github.io/kinitiras-doc)

File: apis/policy/v1alpha1/overridepolicy_types.go [hunk 3]
```diff
@@ -86,6 +90,270 @@ type Overriders struct {
...
+	// K8s means refer another object from current cluster.
+	// +optional
+	K8s *ResourceSelector `json:"k8s,omitempty"`
+	// Http means refer data from remote api.
+	// +optional
+	Http *HttpDataRef `json:"http,omitempty"`
+}
+
+// HttpDataRef defines a http request essential params
+type HttpDataRef struct {
```
Comment (likakuli) [inline apis/policy/v1alpha1/overridepolicy_types.go:211]: not support https for now?

File: go.mod [hunk 4]
```diff
@@ -10,13 +10,13 @@ require (
...
 	golang.org/x/term v0.0.0-20210615171337-6886f2dfbf5b
-	k8s.io/api v0.23.0
+	k8s.io/api v0.23.6
 	k8s.io/apiextensions-apiserver v0.23.0
-	k8s.io/apimachinery v0.23.0
-	k8s.io/client-go v0.23.0
-	k8s.io/code-generator v0.23.6
+	k8s.io/apimachinery v0.23.6
+	k8s.io/client-go v0.23.6
+	k8s.io/code-generator v0.23.0
```
Comment (likakuli) [inline go.mod]: why fallback to v0.23.0?

**Review by likakuli — CHANGES_REQUESTED** [review]

File: apis/policy/v1alpha1/overridepolicy_types.go (hunk 3 above)
Comment (yusank) [reply apis/policy/v1alpha1/overridepolicy_types.go:211]: There is no enough time to support `https` and `http auth`

File: utils/cue/params.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+		if err != nil {
+			return nil, err
+		}
+		refNs, err := parseAndGetRefValue(rs.Namespace, obj)
+		if err != nil {
+			return nil, err
+		}
+
+		klog.V(4).InfoS("get owner reference", "apiVersion", rs.APIVersion, "kind", rs.Kind, "name", refName)
+		obj, err := rc.Namespace(refNs).Get(ctx, refName, metav1.GetOptions{})
```
Comment (likakuli) [inline utils/cue/params.go]: There are performance problem when too many policy using remote resource. Can keep like this this time but need to do this next time.

File: utils/cue/params.go [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+			return nil, err
+		}
+		// use selector
+		s, err := metav1.LabelSelectorAsSelector(handled)
+		if err != nil {
+			return nil, err
+		}
+
+		klog.V(4).InfoS("get object", "label", s.String())
+		list, err := rc.Namespace(rs.Namespace).List(ctx, metav1.ListOptions{
```
Comment (likakuli) [inline utils/cue/params.go]: same as above

File: utils/cue/params.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+	gvk := schema.FromAPIVersionAndKind(or.APIVersion, or.Kind)
+	klog.V(4).InfoS("get owner reference", "apiVersion", or.APIVersion, "kind", or.Kind, "name", or.Name)
+
+	rc, err := c.GetResourceClientByGVK(gvk)
+	if err != nil {
+		klog.ErrorS(err, "GetGroupVersionResource got error", "apiVersion", or.APIVersion, "kind", or.Kind, "name", or.Name)
+		return nil, err
+	}
+
+	return rc.Namespace(obj.GetNamespace()).Get(ctx, or.Name, metav1.GetOptions{})
```
Comment (likakuli) [inline utils/cue/params.go]: same as above

File: apis/policy/v1alpha1/overridepolicy_types.go [hunk 8]
```diff
@@ -86,6 +90,270 @@ type Overriders struct {
...
+	// Type represents current rule operate field type.
+	// +kubebuilder:validation:Enum=annotations;labels;resources;resourcesOversell;tolerations;affinity
+	// +required
+	Type RuleType `json:"type,omitempty"`
+	// Operation represents current operation type.
+	// +kubebuilder:validation:Enum=add;replace;remove
+	// +required
+	Operation OverriderOperator `json:"operation,omitempty"`
+	// Path is field path of current object(e.g. `/spec/affinity`)
+	// If current type is annotations of labels, then only need to provide the key, no need whole path.
```
Comment (likakuli) [inline apis/policy/v1alpha1/overridepolicy_types.go]: annotations of labels, of to or

**Review by likakuli — CHANGES_REQUESTED** [review]

File: utils/cue/params.go (hunk 5 above)
Comment (yusank) [reply utils/cue/params.go]: working on it.

**Pull request merged**